executor/src/witgen/jit/quadratic_symbolic_expression.rs

```diff
@@ -406,6 +417,54 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display> QuadraticSymbolicExp
         };
         Some(Effect::RangeConstraint(solve_for.clone(), constraint))
     }
+
+    /// Tries to detect a range constraint of the form `(X - value1) * (X - value2) = 0`.
+    /// The most common example would be a bit constraint.
+    /// In that case, we can return a range constraint for `X` and mark the
+    /// constraint as complete.
+    // TODO: This function would not detect some equivalent constraints, like
+    // `X * X - X`.
+    // This could be fixed by finding a canonical form for the quadratic
+    // expression, and normalizing the constants.
+    fn try_detect_range_constraint(&self) -> Option<ProcessResult<T, V>> {
+        if !self.linear.is_empty() || !self.constant.is_known_zero() || self.quadratic.len() != 1 {
+            return None;
+        }
+
+        let (factor1, factor2) = &self.quadratic[0];
+        let (var1, value1) = factor1.try_solve_constant_assignment()?;
+        let (var2, value2) = factor2.try_solve_constant_assignment()?;
+
+        if var1 == var2 {
+            let range_constraint = RangeConstraint::from_value(value1)
```

Comment:
Hm, now I'm wondering. If we get a range constraint from the solving, returning a range constraint disjunction is correct even in the case we don't get a full assignment. This could be the case in e.g.
```
(X - Y + 3) * (X - Y + 4) = 0
```
Assume we haven't solved `Y` yet (i.e. it's "unknown") but we know that it is byte-constrained. In this case, we might also learn a new constraint for `X`. This could even work recursively.

So to do that, we would just call `factor1.solve()` and `factor1.solve()` unconditionally, and then go through all pairs of effects and see if we get an effect on the same variable.